config/initializers/errors.rb

```diff
@@ -12,5 +12,37 @@ def initialize(message="Invalid SQL query for this Live Query")
         super
       end
     end
+
+    class ExpectedError < StandardError
+      attr_reader :error_code, :status, :message
+
+      def initialize(error_code, status, message)
+        @error_code = error_code
+        @status = status
+        @message = message
+      end
+
+      def display_error()
+        recursively_print(self)
+      end
+
+      def recursively_print(error, margin="")
+        FOREST_LOGGER.info "#{margin}#{error.message}"
+
+        # NOTICE: Ruby < 2.1.0 doesn't have `cause`
+        if error.respond_to?(:cause) && !error.cause.nil?
+          FOREST_LOGGER.info "#{margin}which was caused by:"
+          recursively_print(error.cause, "#{margin} ")
+        else
+          FOREST_LOGGER.debug error.backtrace.join("\n")
+        end
+      end
+    end
+
+    class HTTP401Error < ExpectedError
+      def initialize(message = "Unauthorized")
+        super(401, :unauthorized, message)
+      end
+    end
   end
 end
```

Comment:
Another case, on TOTP token submit if Forest Server is down. Here is the logs we have:
<img width="1119" alt="screen shot 2018-09-07 at 11 14 59" src="https://user-images.githubusercontent.com/664261/45210279-68049b80-b28f-11e8-9f65-5df176168fc6.png">

The technical error is not really interesting, the only thing we should say is that the service didn't work because Forest looks to be down or in maintenance. Providing traces will make the developer find the reason for the crash.
But he/she will work for nothing.

Reply:
Yes, in this case I agree, I will improve this.